File: common/src/main/java/dev/mayaqq/estrogen/resources/BreastArmorData.java
package dev.mayaqq.estrogen.resources;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.mojang.datafixers.util.Pair;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.GsonHelper;

public class BreastArmorData {
    public ResourceLocation textureLocation;
    public ResourceLocation overlayLocation = null;
    public Pair<Integer, Integer> uv;
    public Pair<Integer, Integer> leftUV;
    public Pair<Integer, Integer> rightUV;
    public Pair<Float, Float> textureSize;

    public BreastArmorData(JsonElement jsonElement) {
        this.textureLocation = ResourceLocation.tryParse(GsonHelper.getAsString(jsonElement.getAsJsonObject(), "texture"));
        this.overlayLocation = jsonElement.getAsJsonObject().has("texture_overlay") ?
                ResourceLocation.tryParse(GsonHelper.getAsString(jsonElement.getAsJsonObject(), "texture_overlay")) : null;
        JsonArray uvArray = GsonHelper.getAsJsonArray(jsonElement.getAsJsonObject(), "uv");
        this.uv = Pair.of(uvArray.get(0).getAsInt(), uvArray.get(1).getAsInt());
        JsonArray leftUVArray = GsonHelper.getAsJsonArray(jsonElement.getAsJsonObject(), "left_uv");
        this.leftUV = Pair.of(leftUVArray.get(0).getAsInt(), leftUVArray.get(1).getAsInt());
        JsonArray rightUVArray = GsonHelper.getAsJsonArray(jsonElement.getAsJsonObject(), "right_uv");
        this.rightUV = Pair.of(rightUVArray.get(0).getAsInt(), rightUVArray.get(1).getAsInt());
        JsonArray textureSizeArray = GsonHelper.getAsJsonArray(jsonElement.getAsJsonObject(), "size");
        this.textureSize = Pair.of(textureSizeArray.get(0).getAsFloat(), textureSizeArray.get(1).getAsFloat());
    }
}

File: src/common/main/kotlin/dev/mayaqq/estrogen/client/features/boobs/data/BreastArmorData.kt
package dev.mayaqq.estrogen.client.features.boobs.data

import com.google.gson.JsonElement
import dev.mayaqq.estrogen.client.content.entityRenderers.boobs.TextureData
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.GsonHelper

data class BreastArmorData(
    val textureLocation: ResourceLocation?,
    val overlayLocation: ResourceLocation?,
    val uv: Pair<Float, Float>,
    val leftUV: Pair<Float, Float>,
    val rightUV: Pair<Float, Float>,
    val textureSize: Pair<Float, Float>
) {

    fun toTextureData(overlay: Boolean): TextureData? {
        return if (overlay) {
            this.toTextureData(overlayLocation)
        } else this.toTextureData(textureLocation)
    }

    fun toTextureData(textureLocation: ResourceLocation?): TextureData? {
        return if (textureLocation == null) null else {
            TextureData(textureLocation, uv.first, uv.second, leftUV.first, leftUV.second, rightUV.first, rightUV.second, textureSize.first, textureSize.second)
        }
    }

    companion object {
        fun fromJson(jsonElement: JsonElement): BreastArmorData {
            val json = jsonElement.asJsonObject
            return BreastArmorData(
                textureLocation = ResourceLocation.tryParse(GsonHelper.getAsString(json, "texture")),
                overlayLocation = json.takeIf { it.has("texture_overlay") }
                    ?.let { ResourceLocation.tryParse(GsonHelper.getAsString(it, "texture_overlay")) },
                uv = GsonHelper.getAsJsonArray(json, "uv").let { 
                    Pair(it[0].asFloat, it[1].asFloat) 
                },
                leftUV = GsonHelper.getAsJsonArray(json, "left_uv").let { 
                    Pair(it[0].asFloat, it[1].asFloat) 
                },
                rightUV = GsonHelper.getAsJsonArray(json, "right_uv").let { 
                    Pair(it[0].asFloat, it[1].asFloat) 
                },
                textureSize = GsonHelper.getAsJsonArray(json, "size").let { 
                    Pair(it[0].asFloat, it[1].asFloat) 
                }
            )
        }
    }
}